### ISSUE TITLE
Support for other bit precisions

### ISSUE THREAD

**Issue by steve3nto** [larq/larq#301]
Is it possible to use larq for 8 bit precision weights and activations?
I like the interface which follows the one of tf.keras, but it seems it is useful only for binary quantization.

Am I missing something? Can I train with 8 bit precision already in Keras or do I need Larq for that?

**Comment by MariaHeuss** [comment]
Hi @steve3nto, thanks for the question. Could you maybe elaborate a bit more on your use case? 

The API (except of the quantizers) of Larq is independent of the bit width so that one can use it with arbitrary precision. As far as I know 8-bit training is not natively supported in Keras. 

We have the DoReFa quantizer in larq, which takes the number of bits that you want to use it with as an argument. So you can use this quantizer for 8 bits, although it is a bit of an odd quantizer, since its quantization range is between 0 and 1. 
We are currently working on adding a more general 8 bit quantizer to larq but in the mean time I would recommend you writing your own quantizer and using it together with the layers in larq. You can also use the tf.quantization.fake_quant_with_min_max_args. If you need more help with that let us know.

**Comment by steve3nto** [comment]
My use case is training a network for semantic segmentation with integer activations and widths to deploy on an embedded device.
I would like to be able to train properly using int8 precision. 

Larq looked very nice for this use case, but indeed the support for 8 bit precision is not as straightforward as training fully binarized models.

How could I write my own larq quantizer? Would it just be a matter of adding a lambda layer to case the data from float32 to int8?

I've never tried tf.quantization.fake_quant_with_min_max_args. How could I use that inside a larq layer?

**Comment by MariaHeuss** [comment]
As far as I understand you want to use both a an input and a kernel quantizer so just adding a lambda layer will not be enough. In the QuantizerBase layer in larq.layers_base the kernel quantizer gets applied to the latent weights inside the call function in each forward pass. The other layers in larq.layers inherit this from the base layer. 
What you could do is to make sure that you import larq.utils as utils and use something like the following as a quantizer in your model. 
```
@utils.register_keras_custom_object
def quantize(x):
    return tf.quantization.fake_quant_with_min_max_args(
         inputs=x, min=-3, max=3, num_bits=8, narrow_range=True
    ) 
```
Alternatively you could write your own quantization function. In this case you would have to make sure that you define a pseudo gradient for the backwards pass since quantization functions are generally not differentiable. As an example for how to do this you can look at the DoReFa quantizer in larq.quantizers.

**Comment by MariaHeuss** [comment]
Did the above approach work for you @steve3nto?

**Comment by MariaHeuss** [comment]
I am closing this due to inactivity. Feel free to reopen if you have more questions.
